src/main/scala/io/iohk/ethereum/vm/utils/Contract.scala

```diff
@@ -0,0 +1,75 @@
+package io.iohk.ethereum.vm.utils
+
+import akka.util.ByteString
+import io.iohk.ethereum.crypto
+import io.iohk.ethereum.crypto._
+import io.iohk.ethereum.domain._
+import io.iohk.ethereum.vm._
+import scala.language.dynamics
+
+case class Contract[W <: WorldStateProxy[W, S], S <: Storage[S]](address: Address, blockHeader: BlockHeader,
+                                                                 world: W, abis: Seq[ABI],
+                                                                 evmConfig: EvmConfig) extends Dynamic {
+
+  def applyDynamic(methodName: String)(args: Any*): ContractMethodCall[W, S] = {
+    callMethod(methodName)(args: _*)
+  }
+
+  def callMethod(methodName: String)(args: Any*): ContractMethodCall[W, S] = {
+    if (methodName.contains("("))
+      callWithSignature(methodName)(args: _*)
+    else
+      callWithMethodName(methodName)(args: _*)
+  }
+
+  private def callWithMethodName(methodName: String)(args: Any*): ContractMethodCall[W, S] = {
+    val matchedAbis = abis
+      .filter { a => a.inputs.size == args.size && a.name == methodName && a.`type` == "function" }
+
+    if (matchedAbis.isEmpty)
+      throw new RuntimeException("No matching ABI found. Please specify the signature")
+    else if (matchedAbis.size > 1)
+      throw new RuntimeException("More than one matching ABI found. Please specify the signature")
+    else {
+      val abi = matchedAbis.head
+      callWithSignature(s"$methodName(${abi.inputs.map(_.`type`).mkString(",")})")(args: _*)
+    }
+  }
+
+  private def callWithSignature(signature: String)(args: Any*): ContractMethodCall[W, S] = {
+    val signatureHash = ByteString(crypto.kec256(signature.getBytes)).take(4)
+    val callData = args.map(parseArg).foldLeft(signatureHash)(_ ++ _)
+    ContractMethodCall(this, callData, blockHeader, evmConfig)
+  }
+
+  private def parseArg(arg: Any): ByteString = arg match {
```

Comment:
Minor detail: Maybe we should make this function public in a companion object as it's a copy from the same function in `EvmTestEnv.scala`?

Reply:
:100: